Enhanced caching proxy, plus performance and reliability improvements

eddyg opened the pull request:
## Summary

This PR introduces comprehensive improvements to caching, error handling, performance, and user experience, including:

- Complete server-side caching proxy replacing CORS workarounds
- Enhanced kiosk mode and scanline rendering
- METAR data augmentation for missing observations
- Content-aware timing and hardware-accelerated scrolling
- Centralized error handling across all modules

## Background

Hi Matt.

I found out about this project recently and was immediately enamored! (I grew up in a household that had TWC on almost any time it wasn't "prime time"!) My dad's birthday was coming up and I knew this would make the cornerstone of the perfect (and unique) gift. So I found a 10" 4:3 LCD monitor with HDMI input on Amazon, ordered a 2GB Raspberry Pi 4, configured it to auto-boot into kiosk mode, and then designed and 3D-printed a case to mount the Pi to the VESA mount on the back of the monitor. (He _loved_ it. 🙂)

While I was tinkering with getting it set up, I started looking in to some occasional "strangeness" with the weather data that I saw in the JavaScript console. One thing led to another, and I had implemented an enhanced proxy solution to provide as much insight as possible into what was happening with the various API calls (and more importantly, _how often_). Then it was adding full _cache_ support to the proxy to ensure I had insight/control into what was actually happening with caching _without_ the browser being involved. Below is an example of the server-side logging showing cache hits, adds, stale cache entries that haven't changed (so there's no need to re-request the full resource), etc. Also notice that the radar images do _not_ have cache headers, but I heuristically cache them as allowed by RFC 7234. (All this makes WS4KP a _much_ nicer "netizen" in terms of API calls, especially on the Mesonet servers! 🎉)
```
🎯 Hit     | https://www.spc.noaa.gov/products/outlook/day2otlk_cat.nolyr.geojson (age: 1199s, remaining: 2401s)
🎯 Hit     | https://www.spc.noaa.gov/products/outlook/day1otlk_cat.nolyr.geojson (age: 1199s, remaining: 2401s)
🎯 Hit     | https://www.spc.noaa.gov/products/outlook/day3otlk_cat.nolyr.geojson (age: 1199s, remaining: 2401s)
📤 Sent    | https://mesonet.agron.iastate.edu/archive/data/2025/06/22/GIS/uscomp/?F=0&P=n0r*.png (no cache directives; not cached)
🎯 Hit     | https://mesonet.agron.iastate.edu/archive/data/2025/06/22/GIS/uscomp/n0r_202506221545.png (age: 1199s, remaining: 2401s)
🎯 Hit     | https://mesonet.agron.iastate.edu/archive/data/2025/06/22/GIS/uscomp/n0r_202506221550.png (age: 1199s, remaining: 2401s)
🎯 Hit     | https://mesonet.agron.iastate.edu/archive/data/2025/06/22/GIS/uscomp/n0r_202506221555.png (age: 600s, remaining: 3000s)
🎯 Hit     | https://mesonet.agron.iastate.edu/archive/data/2025/06/22/GIS/uscomp/n0r_202506221600.png (age: 599s, remaining: 3001s)
🌐 Add     | https://api.weather.gov/gridpoints/IND/83,92/stations?limit=1 (explicit 120s TTL, expires: 2025-06-22T16:15:40.570Z)
🔄 Unchg   | https://api.weather.gov/stations/KMKE/observations/latest (got 304 Not Modified; refreshing cache expiry by 300s)
🌐 Add     | https://mesonet.agron.iastate.edu/archive/data/2025/06/22/GIS/uscomp/n0r_202506221610.png (heuristic 3600s TTL, expires: 2025-06-22T17:13:40.656Z)
```

After that, it snowballed in to more and more changes as I ran into the occasional error, or thought _"it might be cool if..."_

## Details of Changes

- Enhance the kiosk mode startup experience so it is more "appliance like" (so when my dad plugged it in on his birthday, it looked "professional")
- Spent a bunch of time optimizing the _scanlines_ feature since I think it adds a lot to the presentation. (The original implementation at 1024x768 resulted in pretty bad banding, _click to enlarge_.)
<table>
<tr>
<td width="50%">

**Before**
![Before](https://github.com/user-attachments/assets/fa1546b0-1146-45f0-a920-af56995f1162)

</td>
<td width="50%">

**After** 
![After](https://github.com/user-attachments/assets/0d4c8975-6eba-4609-9bfc-a3259a14a75e)

</td>
</tr>
</table>

- Wanted to customize the "travel forecast" with cities important to my dad. I didn't like the complexity of manually creating the `.js` versions, so found a way to use the `.json` version
- Wanted a way to dynamically enable/disable debug logs, so added a module for that
- Needed a way to see details of transition times, so added logging to `weatherdisplay.mjs`
- Saw that I could make some observations "valid" by augmenting them with data present in the METAR, so added the dependency-free `metar-taf-parser` module to the build
- I added a utility function to augment missing weather observation data from the METAR
- Then updated `currentweather.mjs` and `latestobservations.mjs` to augment their from the METAR when possible
- Started moving modules to use centralized error handling so all network errors would be handled consistently, only "bubbling up" unexpected errors
- Added a module to remove expired forecast periods from "stale" API responses
- Updated `extendedforecast.mjs` to skip expired forecasts, changed it to focus on full days, and fixed a day/night temperature pairing issue
- Updated `localforecast.mjs`. In addition to skipping expired forecasts, I added a content-aware transition timing system that uses a DOM-based measurement system to determine the number of lines in each forecast and adjust the dwell time depending on how "full" the screen is.
- Improved the processing and error handling in `spc-outlook.mjs`
- Improved error handling in `regionalforecast.mjs` and `utils/weather.mjs`
- Worked on the weather radar files: optimized radar API usage by only fetching yesterday's data when needed (the Apache index file does not any have cache headers so it's never cached), use centralized URL rewriting for caching proxy support, added debug logging throughout radar processing pipeline, improve canvas context validation and error recovery, and tried to handle worker errors gracefully (by skipping the animation)
- Starting looking into scroll performance and flexibility. I added `utils/scroll-timing.mjs` and worked to replace "magic numbers" with seconds-based timing constants. Switched from `scrollTo` to hardware-accelerated `transform`. Added scroll caching to prevent repeated  DOM queries (which can trigger layout recalculations each time these properties are accessed) on every scroll cycle. Fixed calculations to allow an _arbitrarily-sized_ list of travel forecast cities instead of "20".
- Did the same thing to `hourly.mjs`, including updating the calculations to support flexible hourly forecast lengths (in the future, there could be a setting to display every-other-hour or something)
- Worked on alerts ("hazards") next (both `hazards.mjs` and `currentweatherscroll.mjs`). Updated timing to use seconds-based timing constants. Switched from `scrollTo` to `transform` and did the scroll-caching thing. Updated the CSS for the horizontal alert scroll to go edge-to-edge. Integrated the global speed settings into horizontal scroll timing. 

There are also some minor housekeeping changes: updating the VS Code settings, adding start/stop commands and updating lint rules in `package.json`, etc.

I realize this is an absolutely **massive** set of changes, but I wanted to share them because I think this project is terrific and wanted to give back. With that being said, I started working from the v5.23.3 tag, not realizing how active this project was. I'm not sure what would be involved in applying all the updates since that tag to this PR, because I didn't want to go down that path if you had no interest in merging this. And if you're not, I completely understand — I'll just use them in my private fork. But if you are, it might be enough to bump up to v6.

_In any case, thanks for WS4KP!_

P.S. next up I'm looking in to adding support to fallback to the legacy `MapClick.php` interface when there's stale observation data from the API since I regularly see the API returning observation data that is not current, or forecast data that has expired periods. (Hence the METAR augmentation and expired forecast filtering features...) Is that just me?!

eddyg commented:
Thanks for taking a look, Matt.

Happy to discuss any "design decisions" with you if you have questions about why I did something.

I had two over-arching goals when I started on this:

1. A "clean" Console while the app was running. Meaning, if the API was working "as intended", there wouldn't be any messages. When there was "anomalous" behavior, I do want to see stuff like this:
```
[Log] Latest Observations for station XXXX were augmented with METAR data for windSpeed, windDirection
[Log] Forecast: Removed expired periods: This Afternoon (ended 2025-06-22T18:00:00-04:00), Tonight (ended 2025-06-23T06:00:00-04:00)
[Log] Latest Observations for station XXXX were augmented with METAR data for temperature
```

2) a caching proxy log that showed that each API request was only put on-the-wire when necessary (I think it's important to be a good "netizen" when it comes to API use).

One problem I ran into (with Safari in particular) was aggressive caching of API responses. I could see in the Network tab that the browser was _not_ sending out new API requests (even though the cache headers indicate a 300s TTL) but instead was returning results from the cache. By being able to control the cache headers in the API responses sent to the browser, I alleviated that problem and got a much better picture of how often API requests were coming from the app.

Thanks again!

eddyg commented:
Hi Matt. I've updated my branch to be up-to-date with the latest commits (v5.25.3).

rmitchellscott commented:
This seems to break static builds and therefore the project's current Docker approach:
```
ws.min.js?_=5.25.3:1  ❌ Fetch failed for https://api.weather.gov/points/39.3615,-104.6020 (Fetch error 404 Not Found while fetching http://localhost:8080/api/points/39.3615,-104.6020)
```

rmitchellscott commented:
@eddyg Breaking static builds seems like a poor direction to go in. There's no hard reason (to my knowledge) for this project to require a running Node backend. For example, the previous Node docker image was 362MB, while the static version is 68MB. You could also run the static site from an S3 bucket, etc. 

This patch got your codebase running statically for me:

```diff
diff --git a/server/scripts/modules/utils/url-rewrite.mjs b/server/scripts/modules/utils/url-rewrite.mjs
  index 949b33a..5e51ead 100644
  --- a/server/scripts/modules/utils/url-rewrite.mjs
  +++ b/server/scripts/modules/utils/url-rewrite.mjs
  @@ -8,6 +8,14 @@ const rewriteUrl = (_url) => {
   	// Handle both string URLs and URL objects
   	const url = typeof _url === 'string' ? new URL(_url) : new URL(_url.toString());

+	// Skip rewriting in static builds (no proxy server available)
+	const isStaticBuild = document.querySelector('script[src*="ws.min.js"]') !== null;
+
+	if (isStaticBuild) {
+		return url;
+	}
+
   	// Rewrite the origin to use local proxy server
   	if (url.origin === 'https://api.weather.gov') {
   		url.protocol = window.location.protocol;

```

Curious if something like this would work to maintain compatibility with both methods of running the project.

eddyg commented:
I've updated my latest commit with `Dockerfile` and `Dockerfile.static`. I also adjusted the logic that adds the `User-Agent` header to avoid CORS issues in "static" mode when `User-Agent` is present.

rmitchellscott commented:
A couple breaking changes with this:
1. `RUN rm -f dist/playlist.json` in the regular Dockerfile breaks the music, that's a workaround specifically for static deployments. 
2. Removing `add_header X-Weatherstar true always;` from nginx.conf breaks music in the static deployment.

I wonder if it'd be better to invert the logic on the `window.WS4KP_STATIC_BUILD` so that the dynamic backend injects `window.WS4KP_SERVER_AVAILABLE` as opposed to requiring users who want to host statically manually change HTML, something like this maybe:

```diff
diff --git a/index.mjs b/index.mjs
index ddabd75..f74d1a7 100644
--- a/index.mjs
+++ b/index.mjs
@@ -131,9 +131,34 @@ if (process.env?.DIST === '1') {
        // 'npm run build' and then 'DIST=1 npm start'
        app.use('/scripts', express.static('./server/scripts', staticOptions));
        app.use('/geoip', geoip);
+       // Serve index.html with server availability flag injected
+       app.get('/', (req, res) => {
+               let html = fs.readFileSync('./dist/index.html', 'utf8');
+
+               // Inject server availability flag
+               html = html.replace('<head>', '<head><script>window.WS4KP_SERVER_AVAILABLE=true;</script>');
+
+               res.send(html);
+       });
+
        app.use('/', express.static('./dist', staticOptions));
 } else {
        // Development mode serves files from the server directory: 'npm start'
+       // Inject server availability flag into HTML responses
+       app.use((req, res, next) => {
+               if (req.path === '/' || req.path === '/index.html') {
+                       const originalSend = res.send;
+                       res.send = function(html) {
+                               if (typeof html === 'string' && html.includes('<head>')) {
+                                       html = html.replace('<head>', '<head><script>window.WS4KP_SERVER_AVAILABLE=true;</script>');
+                               }
+                               originalSend.call(this, html);
+                       };
+               }
+               next();
+       });
+
        app.get('/index.html', index);
        app.use('/geoip', geoip);
        app.use('/resources', express.static('./server/scripts/modules'));
diff --git a/server/scripts/modules/utils/fetch.mjs b/server/scripts/modules/utils/fetch.mjs
index 5ba1f8c..cd347c4 100644
--- a/server/scripts/modules/utils/fetch.mjs
+++ b/server/scripts/modules/utils/fetch.mjs
@@ -86,7 +86,7 @@ const fetchAsync = async (_url, responseType, _params = {}) => {
        // User-Agent handling:
        // - Server mode (with caching proxy): Add User-Agent for all requests except excluded hosts
        // - Static mode (direct requests): Only add User-Agent for api.weather.gov, avoiding CORS preflight issues with other services
-       const shouldAddUserAgent = !shouldExcludeUserAgent && (!window.WS4KP_STATIC_BUILD || _url.toString().match(/api\.weather\.gov/));
+       const shouldAddUserAgent = !shouldExcludeUserAgent && (window.WS4KP_SERVER_AVAILABLE || _url.toString().match(/api\.weather\.gov/));
        if (shouldAddUserAgent) {
                headers['user-agent'] = 'Weatherstar 4000+; weatherstar@netbymatt.com';
        }
diff --git a/server/scripts/modules/utils/url-rewrite.mjs b/server/scripts/modules/utils/url-rewrite.mjs
index 17bb7d5..7103720 100644
--- a/server/scripts/modules/utils/url-rewrite.mjs
+++ b/server/scripts/modules/utils/url-rewrite.mjs
@@ -8,8 +8,8 @@ const rewriteUrl = (_url) => {
        // Handle both string URLs and URL objects
        const url = typeof _url === 'string' ? new URL(_url) : new URL(_url.toString());

-       if (window.WS4KP_STATIC_BUILD) {
+       if (!window.WS4KP_SERVER_AVAILABLE) {
                return url;
        }
```

rmitchellscott commented:
> Hmm, what if a "server user" (as opposed to a "static user") wants to bind-mount their own music directory? 🤔

Music bind mounts should happen at `/server/music` and if there are tracks there, playlist.json will get dynamically generated, else it falls back to the build-time `dist/playlist.json` which contains the default tracks.

> This header should be present for both deployment methods via the change in index.mjs... is that not working as expected?

This is Express middleware. Express is not running when these files are served statically, there's no Node.js server to execute server-side Javascript, headers have to be set by the web server (nginx in the case of our static Docker setup).

eddyg commented:
Thanks for the explanation! Makes perfect sense now that you've explained it. Appreciate the additional context.

I've pushed another variation for review. Instead of HTML manipulation, it uses a (cleaner, IMO) template-based
approach, which eliminates middleware overhead when running via `npm start` while maintaining the functionality for detecting static vs. server deployment modes.

(Thanks so much to everyone for all the dialog and feedback — this is the first "fun" project I've had a chance to work on in a long time, and I've really enjoyed it.)

netbymatt commented:
> I think we'll need some guidance from @netbymatt regarding project direction:
> 
> * should the project continue to support both dynamic and static deployment options?
> * should we offer a Docker image for each?
> * which should be the default? (`netbymatt/ws4kp` is dynamic, and then `netbymatt/ws4kp:static` as an option?)

Let me discuss the history of the project and then provide my thoughts on deployment methods.

## Background

I've always intended this to be used as static pages. The use of .ejs came about as a way to organize the html into "modules" that were roughly one-to-one with the javascript .mjs code. Serving it through express was just a convenience so I didn't have to go through a build process every time I changed html, and to clean up some problems with fetch when trying to run from a local file ("c:\ws4kp\index.html" in the address bar).

Then as I stated to use api.weather.gov, 2-3 years ago, I found that there were inconsistent CORS responses, especially in error cases. Thus proxying through express was added. But I always deliberately set up the express routes to exactly match behaviors in cloudfront so I could also proxy api.weather.gov on the live site, but effectively only serve static files as far as the ws4kp code (not data).

The CORS issues I was seeing have largely been resolved with api.weather.gov and when I started seeing the bandwidth costs go up significantly when this app hit a few popular social media and tech web sites in the last month I quickly moved away from proxying and going direct to the api.

For the entire history of api.weather.gov the documentation has stated that an api key will be implemented at some point. But this is still a long way off. The need for an API key would essentially force proxying for all of those calls, and I was prepared to further dial in the express routes to handle this and/or update the cloudfront configuration to apply the appropriate headers when the time came.

I've given further thought to proxying from a caching standpoint and studied some with cloudfront logs. What I've found is that because users are geographically distributed in to 2.5x2.5km grid squares to use the api, there's not a lot of aggregation that goes on. Even the most popular grid squares were getting tens of hits per hour and most grid-related content was unique. Stations/observations were slightly more aggregated because they are spread out more geographically and this may benefit from some caching. But also, api.weather.gov is delivered by akami, which I'm sure gets much more traffic and much better aggregation of gridpoints. I do agree that running your own proxy is a good way to analyze the health and spot bottlenecks in the API calls, I just don't think this repo will scale to that kind of popularity.

## Conclusions

I feel that docker should provide a static, nginx configuration for now... to be revisited if api.weather.gov ever requires api keys. I'm always a fan of small downloads, short, concise code and general programming efficiency. Those wanting to host dynamically can run via node and npm. The bar for that is very low, and while I haven't checked in a while, this probably runs all the way back to node 14 without any fuss.

netbymatt commented:
> Can you explain more about why you wouldn't want to remove the duplicated `.js` versions (`server/scripts/data/{regionalcities,stations,travelcities}.js`) of the data in `datagenerators/output`?

You're right, there's no need for two copies of this data. I still want to keep it separate from the index.html file, however, so that can be small and quick to load and make use of the query string _ (underscore) parameter for updating other resources or letting them load from cache. I think the right solution here is to add them as a build task and have an express route build it on the fly from the .js files. I'll work on this one.

> So that explains the _radar-no-worker_ branch I saw... I'm an iOS user, so let me know if I can be of any help.

Load up that branch and let me know if it's crashing mobile Safari. There's a pretty detailed description of the problem in #74.

eddyg commented:
Hi again Matt. I now see the negative implications of my initial approach of injecting the data into the HTML.

In my most recent commit to this PR, I took the liberty of implementing a solution that removes the large JSON blobs from the EJS templates and implemented a client-side loader utility. This reduces HTML payload size and enables better caching strategies for both server and static deployment modes. 🎉

I appreciate you pointing out the problem with my previous attempt.

eddyg commented:
I took a stab at updating the README to try and add some clarity.

netbymatt commented:
Since this is a significant change in both build/serve methodology and code structure I'm going to set up some infrastructure for a staging site so this can be tested live. As part of this you'll see this pulled into a new branch so I can apply any adjustments as needed.

eddyg commented:
Hey Matt, there's a subtle bug in `radar.mjs` that prevents the radar from animating (at least in Safari):

```
-               this.elem.querySelector('.scroll-area').style.top = `${-this.screenIndex * actualFrameHeight} px`;
+               this.elem.querySelector('.scroll-area').style.top = `${-this.screenIndex * actualFrameHeight}px`;
```

eddyg commented:
Greetings. I've added another commit (plus one for the CSS changes) with some more refactoring to try and improve cross-browser rendering:

- Replace CSS `zoom` with CSS `transform` scaling for better mobile compatibility (`zoom` is a non-standard CSS property that was originally a Microsoft/IE extension. While modern browsers still support it, `transform: scale()` is part of the CSS standard and has much better cross-browser support)
- Add a wrapper-based scaling approach that includes both content and navigation bar
- Replace Almanac layout with CSS Grid for better cross-browser rendering
- Reduce complexity of JavaScript scanline configuration and remove scanline scaling debugging functions
- Allow user to override automatic scanline pick in case it doesn't choose correctly for their particular display
- Update scanline SCSS to use repeating-linear-gradient for better performance
- Refactor settings module to initialize settings upfront and improve change handler declarations
- Improve Current Conditions icon and wind gusts display

Here are some comparison screenshots that I took, for reference:
<table>
<tr>
<td width="20%"></td>
<td width="40%">Before</td>
<td width="40%">After</td>
</tr>

<tr>
<td>
Icon could overflow into wind
</td>
<td>

![Rain Before](https://github.com/user-attachments/assets/522d6dff-ed2d-4856-9996-c85875dbce90)

</td>
<td>

![Rain After](https://github.com/user-attachments/assets/7b7620e9-1b7b-4cc4-b107-4beb02edec8c)

</td>
</tr>

<tr>
<td>
Icon could get reduced in size
</td>
<td>

![Thunderstorms Before](https://github.com/user-attachments/assets/f6612ad2-5091-428b-88d9-dc35e293c7d9)

</td>
<td>

![Thunderstorms After](https://github.com/user-attachments/assets/7819233e-68a2-44f9-b2c9-bbaa1cbd10dd)

</td>
</tr>

<tr>
<td>
Index screen no longer overflows
</td>
<td>

![iPad Index before](https://github.com/user-attachments/assets/8e1650dd-3bab-4334-9f7e-3192df420a99)

</td>
<td>

![iPad index after](https://github.com/user-attachments/assets/28dfeaf5-bcd8-4189-ac66-ac0f9dc11f6e)

</td>
</tr>

<tr>
<td>
Almanac renders properly on iPad/iPhone
</td>
<td>

![Almanac Before](https://github.com/user-attachments/assets/03ddb6b2-671d-46a4-a8af-1fdabe475905)

</td>
<td>

![Almanac After](https://github.com/user-attachments/assets/786ca820-e9d8-4bb5-93dd-a54ede9700ab)

</td>
</tr>

<tr>
<td>
iPad Forecast renders properly in Kiosk mode
</td>
<td>

![Forecast before](https://github.com/user-attachments/assets/6414e04c-7062-4629-ae90-0d2b3ab166f0)

</td>

<td>

_(apparently I didn't take the after screenshot...)_

</td>
</tr>

</table>

As always, I welcome any feedback.

eddyg commented:
Thanks for insight on things to check for and test!

I tested the radar in kiosk mode with Safari, Chrome and Firefox on macOS and all three browsers continued to play the animation even while aggressively resizing the browser. Here's a sample screen recording I made using Safari on high-DPI (retina) display (resized so it would fit within the 10MB GitHub attachment limit): 

![Radar resize test](https://github.com/user-attachments/assets/3fdcfac9-7f3b-4d3d-b4f9-ddf39c342dac)

The other browsers all behaved similarly. My guess is that since the radar rendering code uses `scrollHeight` to measure the actual rendered frame height, and uses CSS `top` positioning for the `.scroll-area`, it adapts to scaling just fine?

I looked at the `Star4000 Large` file (using `fontforge`) and it definitely has issues... like that the Cap Height is 2177, which is actually larger than the Em Size of 2048:
```
Font Metrics for Star4000 Large:
  Em Size: 2048
  Ascent: 1638
  Descent: 410
  Cap Height: 2177.0
  X Height: 1683.0
  Line Gap: 67
  Typo Ascender: 1491
  Typo Descender: -431
  Typo Line Gap: 307
  Win Ascent: 2526
  Win Descent: 530
  HHead Ascent: 1854
  HHead Descent: -434
```
But _all_ the fonts have metrics tables consistency issues, and the culprit _could_ be these variations. Windows likely uses the **Win** metrics, macOS uses the **HHead** metrics, and FreeType-based systems use **Typo** metrics... all of which differ. `Star4000 Large` has the biggest variation, with ratios (as a percentage of the 2048 Em Size) that vary like this:
- **Win** Ascent: 123.3%
- **HHead** Ascent: 90.5%
- **Typo** Ascent: 72.8%

## Star4000 Large.woff:
- Ascent values: HHead=1854, Win=2526, Typo=1491
- Descent values: HHead=434, Win=530, Typo=431
- LineGap values: HHead=67, Typo=307

## Star4000 / Extended / Small.woff:
- Ascent values: HHead=1854, Win=1854, Typo=1491
- Descent values: HHead=434, Win=434, Typo=431
- LineGap values: HHead=67, Typo=307

(Are the problems with `Star4000 Large` the reason there are both `.ttf` and `.woff` versions in the **fonts** directory?)

I processed the **Win** and **Typo** metrics to be consistent with the **HHead** metrics and created new **CM** (Consistent Metrics) variants of all four fonts. I've pushed them to a **consistent-metrics-fonts** branch in my fork. I'd be curious to see if the `font-comparison.html` file shows any difference in rendering on Linux and Windows (in Chrome and Firefox) to know if the fixed metrics make any difference to how the browser renders the fonts. For reference, this is what the comparison looks like in Safari:
![Screenshot 2025-06-30 at 23 15 57@2x](https://github.com/user-attachments/assets/e4ab65b2-49d8-40db-a2ee-b76d0ae7df66)

eddyg commented:
I've tested Safari, Chrome and Firefox on macOS, as well as Mobile Safari on iPhone and iPad, and I feel pretty good about what I see in portrait/landscape/kiosk/fullscreen/etc. in terms of centering, resizing, orientation changes, etc. I'd be very interested in hearing about experiences on other OSs and mobile devices!

For reference, here are screenshots in kiosk mode from an iPad mini and Pro in portrait and landscape showing the scaling, centering, scanlines, etc. as well as from an iPhone, launched as a "Home Screen app", so there is no browser UI:

<table>
<tr>
<td width="10%"></td>
<td width="30%">iPad mini</td>
<td width="30%">iPad Pro</td>
<td width="30%">iPhone (as Home Screen app)</td>
</tr>

<tr>
<td>
Portrait
</td>
<td>

![Screenshot 2025-07-01 at 22 44 39](https://github.com/user-attachments/assets/23e4caa6-fd6e-4273-8877-b5f54a84e2ee)

</td>
<td>

![Screenshot 2025-07-01 at 22 48 04](https://github.com/user-attachments/assets/df34ce2b-088f-4319-a909-83cf2cbc0fd6)

</td>
<td>

![Screenshot 2025-07-02 at 10 19 37](https://github.com/user-attachments/assets/d7034dc7-776a-4706-a2d5-f990d7c6e355)

</td>
</tr>
<tr>
<td>
Landscape
</td>
<td>

![Screenshot 2025-07-01 at 22 45 25](https://github.com/user-attachments/assets/0153be3e-588e-422e-b258-393ee1e52a76)

</td>
<td>

![Screenshot 2025-07-01 at 22 48 17](https://github.com/user-attachments/assets/8c9ebc13-95bc-4865-87cc-2d4e4f0833a9)

</td>
<td>

![Screenshot 2025-07-02 at 10 19 43](https://github.com/user-attachments/assets/c37f5c80-a00e-41a8-8588-7538d089068e)

</td>
</tr>
</table>

For what it's worth, here's a _(reduced in size to be able to attach it)_ screen recording that shows the creation of a Home Screen app on iOS/iPadOS. At the end, I'm force-quitting the "app" and relaunching it:

![Add to Home Screen](https://github.com/user-attachments/assets/402aa15d-e0d3-477e-8da3-fc7e5a52276e)

netbymatt commented:
The screenshots all look good for centering/orientation.

~~I'm curious about the use case for sticky kiosk. I've deliberately left kiosk as not sticky, and forced a opt-in via the url parameter since there was no way out (I see you added ctrl-k to exit). It was my way of ensuring the user knows a little about what they're doing and won't get themselves stuck.~~ Seeing the later readme changes. For other apps I've used a service worker and manifest.json file to create an app. Isn't that the preferred method to allow the user to make a home screen shortcut (or "install" depending on OS)? It might be time for me to go down this path.

I'm very appreciative of all the work you're putting in to this. This will go out as a 6.0 release in the near future.

eddyg commented:
Maybe the README should make the purpose of "Sticky Kiosk" clearer to show that there is a benefit to both iOS and Android users (or at least I think there is?):

> **Sticky Kiosk:** When enabled, stores the kiosk mode preference in local storage so the page automatically enters kiosk mode (maximizing the size of the main weather display without any settings) on subsequent visits. This feature is designed primarily for **iPhone and iPad users** who want to create a Home Screen app experience, since Mobile Safari doesn't support PWA installation via `manifest.json` or the Fullscreen API:

and then, under the steps for Mobile Safari add this:

**For Android and Desktop browsers:** The included `manifest.json` file enables PWA (Progressive Web App) installation. To get the best app-like experience:

1. Configure all your settings first (ignore the "Kiosk" and "Sticky Kiosk" settings)
2. Create a permalink using the "Copy Permalink" feature and manually add `&kiosk=true` to the end
3. Open the edited permalink URL in your browser
4. Look for browser prompts to "Install" or "Add to Home Screen" from the kiosk-enabled URL
5. The PWA will launch directly into kiosk mode (without forcing kiosk mode when accessed from the browser)

Since Mobile Safari doesn't offer the "fullscreen" API, the _only_ way (that I'm aware of) to hide Mobile Safari's UI is by adding it as a Home Screen shortcut. When you do this, _any query parameters are stripped_, and there is no way for the user to manually edit the URL stored with the shortcut. When you launch the shortcut, it runs in an isolated/sandboxed environment with its own localStorage, cookies, etc. that are _not_ shared with the Mobile Safari app. That's why I came up with the "Sticky Kiosk" setting/hack. (It's my understanding that in Android localStorage _is_ shared, so storing `kiosk: true` in the settings would affect both the shortcut _and_ opening WS4KP in the regular browser!)

While service workers and enhanced PWA features could be nice additions, I _think_ the steps above show how to achieve an app-like experience on both mobile platforms as things stand today (I don't have an Android device to confirm that though!)

I'm beyond excited to hear that these changes will make it upstream! 🎉 I've enjoyed working on this. (Note that I will have another commit to add soon to improve the "separation of concerns" when running in "server mode".)

Also, have you had a chance to look at the "Consistent Metrics" fonts to see if they show an alignment difference under various browsers on Windows and/or Linux? If they do, I could regenerate them with the original font names and include them in the PR.

netbymatt commented:
I'm looking at this commit [Centralize icon URL parsing; improve icon error handling](https://github.com/netbymatt/ws4kp/pull/118/commits/804d9e9e33f4b1a976d4625832a77c8cdb0dc073)

I'm not sure about the fallback to clear skies. A deliberate missing icon could prompt a user to file an issue on the repo. It's happened a few times.

The other thing I had considered, because my current logs contain an occasional reference to /images/false, is to make up a known but non-existent image location like `/images/error/<conditionIcon>/<afterLastSlash>`. This lets both the client and server have a 404 in their logs, and the information on the server log can be used to resolve the missing icon.

eddyg commented:
Thanks for the font comparison!

It's interesting that "Linux Chrome" is scaled down compared to the others. 🤔 

Even though there are inconsistencies in the metrics of the other fonts (using the "HHead" metrics as the reference), it must be "Win" metrics that are causing the problem (even though the "Typo" metrics are substantially different from the "HHead" metrics), as it is the one that is off significantly in the **Large** font variant: 

|        |  Ascent HHead | Ascent Win | Ascent Typo | Descent HHead | Descent Win | Descent Typo | LineGap HHead | LineGap Typo  |
| ----: | ----------: | ----------: |  ---------: | ----------: | ---------: | ---------: | ---------: | ---------:  |
| Star4000 |  1854 | 1854 | 1491 | 434 | 434 | 431 | 67 | 307 |
| Star4000 Large  | 1854 | **2526** | 1491 | 434 | **530** | 431 | 67 | 307 |
| Star4000 Extended |  1854 | 1854 | 1491 | 434 | 434 | 431 | 67 | 307 |
| Star4000 Small |  1854 | 1854 | 1491 | 434 | 434 | 431 | 67 | 307 |


So how would you prefer to handle this?

1. Replace all four fonts, keeping the same name
2. Replace all four fonts, use the "CM" suffix, and update all the `.scss` files to reference the new names
3. Replace only **Star4000 Large.woff** with a fixed version, keeping the same name
4. Replace only **Start4000 Large.woff" with the "CM" suffix version, and update all the `.scss` files

I was actually thinking the same thing having a fallback icon. I was going to ask if you be OK will adding something like this `Question-Mark.gif` I created to the `current-conditions` to use as a fallback?

![Question-Mark](https://github.com/user-attachments/assets/8221882c-1cb5-4b8f-9039-ab70738aaeee)

If so, I was going to suggest making `largeIcon()` do something like this:

```
		console.warn(`largeIcon: ${error.message}`);
		return addPath(`Question-Mark.gif?reason=parse-error`);
```

and:

```
			console.warn(`Unknown weather condition '${conditionIcon}' from ${link}; using fallback icon`);
			return addPath(`Question-Mark.gif?reason=${conditionIcon}-missing`);
```

So that errors would be logged in the console and by virtue of the (unused) query parameter on the server side.

netbymatt commented:
## Fonts
Option 4 is the quick fix. But I'm not at all opposed to going to .woff2 format for all the fonts. The .ttf can be removed, but I'll leave it on the server for a while. The logs show requests to resources that I have long-ago retired and to be nice I don't want to break things for those users where possible. That's also why I'm leaning towards the different file names for fonts.

## Icons
There's a "missing" icon in the works https://github.com/netbymatt/ws4kp/issues/94#issuecomment-2956954697. Let's use that one and make sure the query string includes both the source API icon info `afterLastSlash` as well as the attempted parsed result if there is one.

eddyg commented:
Hi Matt. I know you mentioned that you didn't want MapClick API fallback functionality for stale API data, but I wanted to give you a glimpse of what that looks like:

```
[Log] Current Conditions for station KORD: using MapClick data (24 minutes old vs. 144 minutes old)
[Log] Latest Observations for station KLOT: using MapClick data (30 minutes old vs. 155 minutes old)
[Log] Latest Observations for station KPWK: using MapClick data (23 minutes old vs. 143 minutes old; provides missing required: windDirection)
[Log] Current Conditions for station KORD: using MapClick data (34 minutes old vs. 154 minutes old)
[Log] Latest Observations for station KUGN: using MapClick data (34 minutes old vs. 154 minutes old)
[Log] Latest Observations for station KLOT: using MapClick data (40 minutes old vs. 165 minutes old)
[Log] Latest Observations for station KJOT: using MapClick data (50 minutes old vs. 150 minutes old)
[Log] Regional Observations for station KDSM: using MapClick data (31 minutes old vs. 151 minutes old)
[Log] Regional Observations for station KDBQ: using MapClick data (32 minutes old vs. 152 minutes old)
[Log] Regional Observations for station KDUH: using MapClick data (30 minutes old vs. 150 minutes old)
[Log] Latest Observations for station KMWS: using MapClick data (43 minutes old vs. 136 minutes old; provides missing required: textDescription)
[Log] Regional Forecast for station KPRB: using MapClick data (13 minutes old vs. 100 minutes old)
```

This is providing _much_ more up-to-date data on the modules where I've integrated it (Current Conditions, Latest Observations, and Regional Observations). I'd say this provides better value than the "METAR augmentation", which I see considerably less often. (I haven't yet looked into the effect it may have on other modules, like Local Forecast.)

The overwhelming majority of code changes are in the MapClick adapter itself. The integration into the first three modules seemed relatively small, which means it would be easy to remove some time down the road once api.weather.gov starts providing consistent, timely data.

If you want to take a look, I've created a stacked PR (based on this PR's branch) in my fork [that you can take a look at here](https://github.com/eddyg/ws4kp/pull/1).

netbymatt commented:
I published the most recent commit including the mapclick stacked PR to the preview site https://preview.weatherstar.netbymatt.com so I could explore this a little more. Mapclick does include the necessary cors headers to that looks to be good to go.

eddyg commented:
I've created [a new stacked PR](https://github.com/eddyg/ws4kp/pull/2) with a `stations.json` post-processing script that improves the "display name" for a large number of NWS stations and resolves any duplicate names to be unique, using a priority system (so an international airport in a given city gets the bare city name, any others that resolve to the same city get some kind of qualifier). The "fallback of last resort" is appending the station ID to provide uniqueness (but this is _extremely_ rare.) My goal was to try and get as many stations to be a _city_, at least as much as that's possible given the information in the name of the station. For international airports, I use an exception table when a station requires external knowledge, e.g. the "Harry Reid International Airport" station, but my goal was to try and keep hard-coded exceptions to a minimum.

Do you have any idea if the NWS updates station names? Or, once they're in use, do they keep the same name? (I'm amazed at the lack of any kind of naming standard... how hard would it be to use a consistent format like "City / Name"?! 😭)

eddyg commented:
> Are the font changes included in what was merged? It appears to be separate branch per my git client. But maybe that's an exploratory branch: `consistent-metrics-fonts`

Correct, they were a separate branch because the work was experimental — I had my suspicions about what was wrong, but needed some testing on other OSs/browsers to see if that's really what it was.

The key thing is that I generated the fonts with the " CM" (Consistent Metrics) suffix for differentiation, which means that any CSS references would need to be updated to use the new names if we used the fonts in that branch.

Based on the test results, the other alternative is to create a new **Star4000 Large** with the _same name_ and simply replace the `.woff` font since it seems to be the only one that actually casual rendering issues. (Could also remove the `.ttf`?)
 
> Does a bit of work need to be done to add the post-processor naming routine into the gulp workflow, as well as my recent station overrides change that was unfortunately added nearly in parallel?

Probably. I saw some commits about station overrides and was like _"what are the chances?"_ I am very busy with `$REALJOB` stuff this week and then will be traveling next week, but can take a look at both of these things when I'm back.

> I hope that your work on the METAR supplement and mapclick fallback didn't go to waste. api.weather.gov just resolved the latency issue.

Ha! Story of my life! It'd be great if the official API is reliable and provides up-to-date data, so this is good news if true!

I just checked the console of the one I keep running (no debug flags), and it's still removing expired forecast periods long after they should have been removed from the API response, so that's one data point:
```
2025-08-05 03:27:28.510 [L] 🚮 Forecast: Removed expired periods: This Afternoon (ended 2025-08-04T18:00:00-04:00)
2025-08-05 07:27:28.034 [L] 🚮 Forecast: Removed expired periods: Overnight (ended 2025-08-05T06:00:00-04:00)
```
This gets logged without any debug flags enabled. I'll enable `&debug=currentweather,latestobservations,mapclick` and see what happens! Thanks for the update!
